app/Pasteboard.m

```diff
@@ -0,0 +1,247 @@
+
+#define INITIAL_BUFFER_CAP 1024
+// 8MB: https://stackoverflow.com/a/3523175
+#define MAXIMAL_BUFFER_CAP 8*1024*1024
+
+// If pasteboard contents were changed since file was opened,
+// all read operations on in return error
+static int check_read_generation(clip_fd *fd) {
+    UIPasteboard *pb = UIPasteboard.generalPasteboard;
+
+    uint64_t pb_gen = (uint64_t) pb.changeCount;
+    uint64_t fd_gen = fd_priv(fd).generation;
+
+    if (fd_gen == 0 || fd->offset == 0) {
+        fd_priv(fd).generation = pb_gen;
+    } else if (fd_gen != pb_gen) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static const char *get_data(clip_fd *fd, size_t *len) {
+    if (fd_priv(fd).buffer != NULL) {
+        *len = fd_priv(fd).buffer_len;
+        return fd_priv(fd).buffer;
+    }
+
+    if (check_read_generation(fd) != 0) {
+        return NULL;
+    }
+
+    NSString *contents = UIPasteboard.generalPasteboard.string;
+    *len = contents.length;
+    return contents.UTF8String;
+}
+
+static int realloc_to_fit(clip_fd* fd, size_t fit_len) {
+    // (Re)allocate buffer if there's not enough space to fit fit_len
+    if (fit_len <= fd_priv(fd).buffer_cap) {
+        return 0;
+    }
+    if (fit_len > MAXIMAL_BUFFER_CAP) {
+        return 1;
+    }
+
+    size_t size = fd_priv(fd).buffer_cap * 2;
+    if (size == 0) {
+        size = INITIAL_BUFFER_CAP;
+    }
+    while (size < fit_len) size *= 2;
+
+    void *new_buf = realloc(fd_priv(fd).buffer, size);
+    if (new_buf == NULL) {
+        return 1;
+    }
+
+    fd_priv(fd).buffer = new_buf;
+    fd_priv(fd).buffer_cap = size;
+
+    return 0;
+}
+
+// buffer => UIPasteboard
+static int clipboard_wsync(clip_fd *fd) {
+    if (fd_priv(fd).buffer == NULL) {
+        return 0;
+    }
+
+    void *data = fd_priv(fd).buffer;
+    size_t len = fd_priv(fd).buffer_len;
+
+    // FIXME(stek29): This logs "Returning local object of class NSString"
+    // and I have no idea why (or how to fix it)
+    UIPasteboard.generalPasteboard.string = [[NSString alloc]
+                                             initWithBytes:data
+                                             length:len
+                                             encoding:NSUTF8StringEncoding];
+
+    // Reset generation since we've just updated UIPasteboard
+    // note: offset doesn't change
+    fd_priv(fd).generation = 0;
+
+    return 0;
+}
+
+// UIPasteboard => buffer, return len
+static ssize_t clipboard_rsync(clip_fd *fd) {
+    if (fd_priv(fd).buffer != NULL) {
+        free(fd_priv(fd).buffer);
+        fd_priv(fd).buffer = NULL;
+        fd_priv(fd).buffer_cap = 0;
+        fd_priv(fd).buffer_len = 0;
+    }
+
+    size_t len;
+    const void *data = get_data(fd, &len);
+    
+    // Make sure size is still INITIAL_BUFFER_CAP based
+    if (realloc_to_fit(fd, len)) {
+        return _ENOMEM;
+    }
+
+    memcpy(fd_priv(fd).buffer, data, len);
+    fd_priv(fd).buffer_len = len;
+
+    return len;
+}
+
+static int clipboard_poll(clip_fd *fd) {
```

Comment:
This should just always return POLL_READ | POLL_WRITE, because reading and writing will never block. (The return from poll is defined as the operations that wouldn't block, not the operations that would succeed.)

Reply:
Agreed. I was thinking of making them blocking at first, but I haven't gotten to it.